File: reference/rocketride-server/packages/client-typescript/src/client/account.ts
/**
 * MIT License
 *
 * Copyright (c) 2026 Aparavi Software AG
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * Account API namespace for the RocketRide TypeScript SDK.
 *
 * Provides typed methods for managing the authenticated user's profile,
 * API keys, organization, members, and teams via DAP commands over the
 * existing WebSocket connection.
 */

import type { RocketRideClient } from './client.js';
import type { ConnectResult } from './types/client.js';
import type { OrgDetail, ApiKeyRecord, MemberRecord, TeamRecord, TeamDetail, ProfileUpdate, CreateKeyParams, InviteMemberParams, TeamMemberParams } from './types/account.js';

// =============================================================================
// ACCOUNT API CLASS
// =============================================================================

/**
 * Typed wrapper around the `rrext_account_*` DAP commands.
 *
 * Accessed via `client.account` — not instantiated directly. All methods
 * delegate to {@link RocketRideClient.dap} which handles envelope
 * unwrapping and error propagation.
 */
export class AccountApi {
	/** @param client - The parent RocketRideClient that owns this namespace. */
	constructor(private client: RocketRideClient) {}

	// =========================================================================
	// PROFILE
	// =========================================================================

	/**
	 * Fetches the current user's profile from the server.
	 *
	 * @returns The user's profile data.
	 */
	async getProfile(): Promise<ConnectResult> {
		return this.client.call<ConnectResult>('rrext_account_me', { subcommand: 'get' });
	}

	/**
	 * Persists updated profile fields.
	 *
	 * @param fields - The profile fields to update.
	 */
	async updateProfile(fields: ProfileUpdate): Promise<void> {
		await this.client.call('rrext_account_me', { subcommand: 'update', ...fields });
	}

	/**
	 * Sets the user's preferred default team.
	 *
	 * @param teamId - The team ID to set as default.
	 */
	async setDefaultTeam(teamId: string): Promise<void> {
		await this.client.call('rrext_account_me', { subcommand: 'set_default_team', teamId });
	}

	/**
	 * Switches the user's active organization.
	 *
	 * The server updates the user's default_org_id and resets the default
	 * team to the first team in the new org. All connections for this user
	 * receive a refreshed AccountInfo via shell:accountUpdate.
	 *
	 * @param orgId - The org ID to switch to.
	 */
	async setDefaultOrg(orgId: string): Promise<void> {
		await this.client.call('rrext_account_me', { subcommand: 'set_default_org', orgId });
	}

	/**
	 * Permanently deletes the current user's account.
	 */
	async deleteAccount(): Promise<void> {
		await this.client.call('rrext_account_me', { subcommand: 'delete' });
	}

	// =========================================================================
	// ORGANIZATION
	// =========================================================================

	/**
	 * Fetches the organization detail for the given org.
	 *
	 * @param orgId - Organisation UUID. The server may infer the org if omitted.
	 * @returns The organization detail (id, name, plan, memberCount, teamCount).
	 */
	async getOrg(orgId?: string): Promise<OrgDetail> {
		return this.client.call<OrgDetail>('rrext_account_org', { subcommand: 'get', ...(orgId ? { orgId } : {}) });
	}

	/**
	 * Updates the organization name.
	 *
	 * @param orgId - Organisation UUID.
	 * @param name  - The new organization name.
	 */
	async updateOrgName(orgId: string, name: string): Promise<void> {
		await this.client.call('rrext_account_org', { subcommand: 'update', orgId, name });
	}

	// =========================================================================
	// API KEYS
	// =========================================================================

	/**
	 * Fetches the list of API keys for the current user.
	 *
	 * @returns Array of API key records.
	 */
	async listKeys(): Promise<ApiKeyRecord[]> {
		const body = await this.client.call('rrext_account_keys', { subcommand: 'list' });
		return body.keys ?? [];
	}

	/**
	 * Creates a new API key and returns the raw key string.
	 *
	 * @param params - Key creation parameters (name, permissions, expiresAt).
	 * @returns Object containing the raw key string.
	 */
	async createKey(params: CreateKeyParams): Promise<{ key: string }> {
		const body = await this.client.call('rrext_account_keys', { subcommand: 'create', ...params });
		return { key: body.key };
	}

	/**
	 * Revokes an API key by its ID.
	 *
	 * @param keyId - The key to revoke.
	 */
	async revokeKey(keyId: string): Promise<void> {
		await this.client.call('rrext_account_keys', { subcommand: 'revoke', keyId });
	}

	// =========================================================================
	// MEMBERS
	// =========================================================================

	/**
	 * Fetches the flat list of organization members.
	 *
	 * @param orgId - Organisation UUID.
	 * @returns Array of member records.
	 */
	async listMembers(orgId: string): Promise<MemberRecord[]> {
		const body = await this.client.call('rrext_account_members', { subcommand: 'list', orgId });
		return body.members ?? [];
	}

	/**
	 * Sends an invitation to a new organization member.
	 *
	 * @param orgId  - Organisation UUID.
	 * @param params - Invitation parameters (email, givenName, familyName, role).
	 */
	async inviteMember(orgId: string, params: InviteMemberParams): Promise<void> {
		await this.client.call('rrext_account_members', { subcommand: 'invite', orgId, ...params });
	}

	/**
	 * Updates an organization member's role.
	 *
	 * @param orgId  - Organisation UUID.
	 * @param userId - The member's user ID.
	 * @param role   - The new role string.
	 */
	async updateMemberRole(orgId: string, userId: string, role: string): Promise<void> {
		await this.client.call('rrext_account_members', { subcommand: 'update', orgId, userId, role });
	}

	/**
	 * Removes an organization member.
	 *
	 * @param orgId  - Organisation UUID.
	 * @param userId - The member's user ID.
	 */
	async removeMember(orgId: string, userId: string): Promise<void> {
		await this.client.call('rrext_account_members', { subcommand: 'delete', orgId, userId });
	}

	/**
	 * Resends the initialization email for a pending org member.
	 *
	 * @param orgId  - Organisation UUID.
	 * @param userId - The pending member's user ID.
	 */
	async resendInvite(orgId: string, userId: string): Promise<void> {
		await this.client.call('rrext_account_members', { subcommand: 'resend_invite', orgId, userId });
	}

	// =========================================================================
	// TEAMS
	// =========================================================================

	/**
	 * Fetches the flat list of teams in the organization.
	 *
	 * @param orgId - Organisation UUID.
	 * @returns Array of team summary records.
	 */
	async listTeams(orgId: string): Promise<TeamRecord[]> {
		const body = await this.client.call('rrext_account_teams', { subcommand: 'list', orgId });
		return body.teams ?? [];
	}

	/**
	 * Fetches full detail (including member list) for a specific team.
	 *
	 * @param orgId  - Organisation UUID.
	 * @param teamId - The team to load.
	 * @returns The team detail with nested members.
	 */
	async getTeamDetail(orgId: string, teamId: string): Promise<TeamDetail> {
		return this.client.call<TeamDetail>('rrext_account_teams', { subcommand: 'get', orgId, teamId });
	}

	/**
	 * Creates a new team.
	 *
	 * @param orgId - Organisation UUID.
	 * @param name  - The team name.
	 */
	async createTeam(orgId: string, name: string): Promise<void> {
		await this.client.call('rrext_account_teams', { subcommand: 'create', orgId, name });
	}

	/**
	 * Deletes a team.
	 *
	 * @param orgId  - Organisation UUID.
	 * @param teamId - The team to delete.
	 */
	async deleteTeam(orgId: string, teamId: string): Promise<void> {
		await this.client.call('rrext_account_teams', { subcommand: 'delete', orgId, teamId });
	}

	/**
	 * Adds a member to a team with specified permissions.
	 *
	 * @param orgId  - Organisation UUID.
	 * @param params - Parameters (teamId, userId, permissions).
	 */
	async addTeamMember(orgId: string, params: TeamMemberParams): Promise<void> {
		await this.client.call('rrext_account_teams', { subcommand: 'add_member', orgId, ...params });
	}

	/**
	 * Updates a team member's permissions.
	 *
	 * @param orgId  - Organisation UUID.
	 * @param params - Parameters (teamId, userId, permissions).
	 */
	async updateTeamMemberPerms(orgId: string, params: TeamMemberParams): Promise<void> {
		await this.client.call('rrext_account_teams', { subcommand: 'update_member', orgId, ...params });
	}

	/**
	 * Removes a member from a team.
	 *
	 * @param orgId  - Organisation UUID.
	 * @param params - Parameters (teamId, userId).
	 */
	async removeTeamMember(orgId: string, params: { teamId: string; userId: string }): Promise<void> {
		await this.client.call('rrext_account_teams', { subcommand: 'delete_member', orgId, ...params });
	}

	// =========================================================================
	// ENVIRONMENT SECRETS
	// =========================================================================

	/**
	 * Returns the available ROCKETRIDE_* key names from the merged environment.
	 * Does not return values — only key names for use in dropdowns.
	 *
	 * @returns Array of key names (e.g. ['ROCKETRIDE_ANTHROPIC_KEY', 'ROCKETRIDE_OPENAI_KEY']).
	 */
	async getEnvironmentKeys(): Promise<string[]> {
		const body = await this.client.call<{ keys: string[] }>('rrext_account_me', { subcommand: 'env_keys' });
		return body.keys ?? [];
	}

	/**
	 * Reads the environment dict for a scope (org, team, or user).
	 *
	 * @param scope   - One of 'org', 'team', 'user'.
	 * @param scopeId - For org: orgId. For team: teamId. For user: omit (uses current user).
	 * @returns Decrypted key-value dict.
	 */
	async getEnv(scope: 'org' | 'team' | 'user', scopeId?: string): Promise<Record<string, string>> {
		const command = scope === 'org' ? 'rrext_account_org' : scope === 'team' ? 'rrext_account_teams' : 'rrext_account_me';
		const args: Record<string, unknown> = { subcommand: 'get_env' };
		if (scope === 'org' && scopeId) args.orgId = scopeId;
		if (scope === 'team' && scopeId) args.teamId = scopeId;
		const body = await this.client.call<{ env: Record<string, string> }>(command, args);
		return body.env ?? {};
	}

	/**
	 * Writes the full environment dict for a scope (org, team, or user).
	 * Replaces the entire set of keys at that scope level.
	 *
	 * @param scope   - One of 'org', 'team', 'user'.
	 * @param env     - Full key-value dict to store.
	 * @param scopeId - For org: orgId. For team: teamId. For user: omit.
	 */
	async setEnv(scope: 'org' | 'team' | 'user', env: Record<string, string>, scopeId?: string): Promise<void> {
		const command = scope === 'org' ? 'rrext_account_org' : scope === 'team' ? 'rrext_account_teams' : 'rrext_account_me';
		const args: Record<string, unknown> = { subcommand: 'set_env', env };
		if (scope === 'org' && scopeId) args.orgId = scopeId;
		if (scope === 'team' && scopeId) args.teamId = scopeId;
		await this.client.call(command, args);
	}
}

File: reference/rocketride-server/packages/client-python/src/rocketride/account.py
# MIT License
#
# Copyright (c) 2026 Aparavi Software AG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
Account API namespace for the RocketRide Python SDK.

Provides typed methods for managing the authenticated user's profile,
API keys, organization, members, and teams via DAP commands over the
existing WebSocket connection.

Usage:
    profile = await client.account.get_profile()
    await client.account.update_profile(displayName='New Name')
    keys = await client.account.list_keys()
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .types.account import (
    AccountProfile,
    ApiKeyRecord,
    CreateKeyResult,
    MemberRecord,
    OrgDetail,
    TeamDetail,
    TeamRecord,
)

if TYPE_CHECKING:
    from .client import RocketRideClient


class AccountApi:
    """
    Account management namespace on RocketRideClient.

    Accessed via ``client.account`` -- not instantiated directly. All methods
    delegate to the parent client's ``call()`` method which handles envelope
    construction, sending, error detection, and tracing.
    """

    def __init__(self, client: RocketRideClient) -> None:
        """
        Bind this namespace to its parent client.

        Args:
            client: The RocketRideClient instance that owns this namespace.
        """
        self._client = client

    # =========================================================================
    # PROFILE
    # =========================================================================

    async def get_profile(self) -> AccountProfile:
        """
        Fetch the current user's profile from the server.

        Returns:
            The user's profile data.
        """
        return await self._client.call('rrext_account_me', subcommand='get')

    async def update_profile(self, **fields: str) -> None:
        """
        Persist updated profile fields.

        Args:
            **fields: Profile fields to update (displayName, givenName, etc.).
                All values are strings; an empty string means no change.
        """
        await self._client.call('rrext_account_me', subcommand='update', **fields)

    async def set_default_team(self, team_id: str) -> None:
        """
        Set the user's preferred default team.

        Args:
            team_id: The team ID to set as default.
        """
        await self._client.call('rrext_account_me', subcommand='set_default_team', teamId=team_id)

    async def delete_account(self) -> None:
        """Permanently delete the current user's account."""
        await self._client.call('rrext_account_me', subcommand='delete')

    # =========================================================================
    # ORGANIZATION
    # =========================================================================

    async def get_org(self, org_id: str) -> OrgDetail:
        """
        Fetch the organization detail for the given org.

        Args:
            org_id: Organisation UUID.

        Returns:
            The organization detail (id, name, plan, memberCount, teamCount).
        """
        return await self._client.call('rrext_account_org', subcommand='get', orgId=org_id)

    async def update_org_name(self, org_id: str, name: str) -> None:
        """
        Update the organization name.

        Args:
            org_id: Organisation UUID.
            name: The new organization name.
        """
        await self._client.call('rrext_account_org', subcommand='update', orgId=org_id, name=name)

    # =========================================================================
    # API KEYS
    # =========================================================================

    async def list_keys(self) -> list[ApiKeyRecord]:
        """
        Fetch the list of API keys for the current user.

        Returns:
            Array of API key records.
        """
        body = await self._client.call('rrext_account_keys', subcommand='list')
        return body.get('keys', [])

    async def create_key(
        self,
        *,
        name: str,
        permissions: list[str] | None = None,
        expires_at: str | None = None,
        team_id: str | None = None,
    ) -> CreateKeyResult:
        """
        Create a new API key (PAT) and return the raw key string.

        When team_id is None (default) the key inherits all teams and
        permissions from the user. When team_id is set, the key is scoped
        to that team and permissions must be provided. Effective permissions
        are always intersected with the user's actual permissions at auth time.

        Args:
            name: Human-readable label for the key.
            permissions: Permission strings. Required when team_id is set;
                None means inherit all from the user.
            expires_at: Optional ISO timestamp for key expiration. None means no expiry.
            team_id: Optional team UUID to scope this key to.

        Returns:
            Dict containing the raw key string under the ``key`` field.
        """
        kwargs: dict = {
            'subcommand': 'create',
            'name': name,
        }
        if permissions is not None:
            kwargs['permissions'] = permissions
        if expires_at is not None:
            kwargs['expiresAt'] = expires_at
        if team_id is not None:
            kwargs['teamId'] = team_id

        body = await self._client.call('rrext_account_keys', **kwargs)
        return {'key': body['key']}

    async def revoke_key(self, key_id: str) -> None:
        """
        Revoke an API key by its ID.

        Args:
            key_id: The key to revoke.
        """
        await self._client.call('rrext_account_keys', subcommand='revoke', keyId=key_id)

    # =========================================================================
    # MEMBERS
    # =========================================================================

    async def list_members(self, org_id: str) -> list[MemberRecord]:
        """
        Fetch the flat list of organization members.

        Args:
            org_id: Organisation UUID.

        Returns:
            Array of member records.
        """
        body = await self._client.call('rrext_account_members', subcommand='list', orgId=org_id)
        return body.get('members', [])

    async def invite_member(
        self,
        org_id: str,
        *,
        email: str,
        given_name: str,
        family_name: str,
        role: str,
    ) -> None:
        """
        Send an invitation to a new organization member.

        Args:
            org_id: Organisation UUID.
            email: Email address of the person to invite.
            given_name: First / given name of the invitee.
            family_name: Last / family name of the invitee.
            role: Organization-level role to assign (e.g. "admin" or "member").
        """
        await self._client.call(
            'rrext_account_members',
            subcommand='invite',
            orgId=org_id,
            email=email,
            givenName=given_name,
            familyName=family_name,
            role=role,
        )

    async def update_member_role(self, org_id: str, user_id: str, role: str) -> None:
        """
        Update an organization member's role.

        Args:
            org_id: Organisation UUID.
            user_id: The member's user ID.
            role: The new role string.
        """
        await self._client.call(
            'rrext_account_members',
            subcommand='update',
            orgId=org_id,
            userId=user_id,
            role=role,
        )

    async def remove_member(self, org_id: str, user_id: str) -> None:
        """
        Remove an organization member.

        Args:
            org_id: Organisation UUID.
            user_id: The member's user ID.
        """
        await self._client.call(
            'rrext_account_members',
            subcommand='delete',
            orgId=org_id,
            userId=user_id,
        )

    # =========================================================================
    # TEAMS
    # =========================================================================

    async def list_teams(self, org_id: str) -> list[TeamRecord]:
        """
        Fetch the flat list of teams in the organization.

        Args:
            org_id: Organisation UUID.

        Returns:
            Array of team summary records.
        """
        body = await self._client.call('rrext_account_teams', subcommand='list', orgId=org_id)
        return body.get('teams', [])

    async def get_team_detail(self, org_id: str, team_id: str) -> TeamDetail:
        """
        Fetch full detail (including member list) for a specific team.

        Args:
            org_id: Organisation UUID.
            team_id: The team to load.

        Returns:
            The team detail with nested members.
        """
        return await self._client.call(
            'rrext_account_teams',
            subcommand='get',
            orgId=org_id,
            teamId=team_id,
        )

    async def create_team(self, org_id: str, name: str) -> None:
        """
        Create a new team.

        Args:
            org_id: Organisation UUID.
            name: The team name.
        """
        await self._client.call('rrext_account_teams', subcommand='create', orgId=org_id, name=name)

    async def delete_team(self, org_id: str, team_id: str) -> None:
        """
        Delete a team.

        Args:
            org_id: Organisation UUID.
            team_id: The team to delete.
        """
        await self._client.call(
            'rrext_account_teams',
            subcommand='delete',
            orgId=org_id,
            teamId=team_id,
        )

    async def add_team_member(
        self,
        org_id: str,
        *,
        team_id: str,
        user_id: str,
        permissions: list[str],
    ) -> None:
        """
        Add a member to a team with specified permissions.

        Args:
            org_id: Organisation UUID.
            team_id: The team to add the member to.
            user_id: The user ID of the member.
            permissions: Permissions to grant within the team.
        """
        await self._client.call(
            'rrext_account_teams',
            subcommand='add_member',
            orgId=org_id,
            teamId=team_id,
            userId=user_id,
            permissions=permissions,
        )

    async def update_team_member_perms(
        self,
        org_id: str,
        *,
        team_id: str,
        user_id: str,
        permissions: list[str],
    ) -> None:
        """
        Update a team member's permissions.

        Args:
            org_id: Organisation UUID.
            team_id: The team containing the member.
            user_id: The user ID of the member.
            permissions: New permissions to set within the team.
        """
        await self._client.call(
            'rrext_account_teams',
            subcommand='update_member',
            orgId=org_id,
            teamId=team_id,
            userId=user_id,
            permissions=permissions,
        )

    async def remove_team_member(
        self,
        org_id: str,
        *,
        team_id: str,
        user_id: str,
    ) -> None:
        """
        Remove a member from a team.

        Args:
            org_id: Organisation UUID.
            team_id: The team to remove the member from.
            user_id: The user ID of the member.
        """
        await self._client.call(
            'rrext_account_teams',
            subcommand='delete_member',
            orgId=org_id,
            teamId=team_id,
            userId=user_id,
        )

    # =========================================================================
    # ENVIRONMENT
    # =========================================================================

    async def get_environment_keys(self) -> list[str]:
        """
        Return the merged list of ROCKETRIDE_* key names (no values).

        The list includes keys from all scopes (org → team → user) merged
        in the same precedence order the server uses for pipeline resolution.

        Returns:
            Sorted list of ROCKETRIDE_* key names.
        """
        body = await self._client.call(
            'rrext_account_me',
            subcommand='env_keys',
        )
        return body.get('keys', [])

    async def get_env(
        self,
        scope: str,
        scope_id: str | None = None,
    ) -> dict[str, str]:
        """
        Read the environment dict for a scope (org, team, or user).

        Args:
            scope: One of 'org', 'team', 'user'.
            scope_id: For org: orgId. For team: teamId. For user: omit.

        Returns:
            Decrypted key-value dict of ROCKETRIDE_* variables.
        """
        if scope == 'org':
            command = 'rrext_account_org'
            kwargs = {'subcommand': 'get_env'}
            if scope_id:
                kwargs['orgId'] = scope_id
        elif scope == 'team':
            command = 'rrext_account_teams'
            kwargs = {'subcommand': 'get_env'}
            if scope_id:
                kwargs['teamId'] = scope_id
        else:
            command = 'rrext_account_me'
            kwargs = {'subcommand': 'get_env'}

        body = await self._client.call(command, **kwargs)
        return body.get('env', {})

    async def set_env(
        self,
        scope: str,
        env: dict[str, str],
        scope_id: str | None = None,
    ) -> None:
        """
        Write the full environment dict for a scope (org, team, or user).

        Replaces the entire set of keys at that scope level.

        Args:
            scope: One of 'org', 'team', 'user'.
            env: Full key-value dict to store.
            scope_id: For org: orgId. For team: teamId. For user: omit.
        """
        if scope == 'org':
            command = 'rrext_account_org'
            kwargs = {'subcommand': 'set_env', 'env': env}
            if scope_id:
                kwargs['orgId'] = scope_id
        elif scope == 'team':
            command = 'rrext_account_teams'
            kwargs = {'subcommand': 'set_env', 'env': env}
            if scope_id:
                kwargs['teamId'] = scope_id
        else:
            command = 'rrext_account_me'
            kwargs = {'subcommand': 'set_env', 'env': env}

        await self._client.call(command, **kwargs)
